07PerformanceProgramming/Sec04ConcurrencyProblems.md

```diff
@@ -0,0 +1,103 @@
+---
+title: Concurrency Problems
+---
+
+## Issues with concurrency
+Running a program with several concurrent threads of execution can encounter several issues that are not present when running a single thread. These problems can include
+
+* Deadlock: the system is locked, not doing any work and not changing state.
+* Livelock: the system is apparently doing _something_, changing state, but no work is being done.
+* Resource Starvation: one or more of the threads is unable to perform work, being unable to access a necessary resource.
+* Race Condition: a thread ensures an action is possible, but this is not longer the case when the action is attempted, due to another thread.
+
+## Dining Philosophers
+A classical computer science analogy to illustrate issues with concurrency and resource contention.
+
+### The set up
+Five hungry yet contemplative philosophers are seated around a table. In front of each of them is a bowl of noodles. In between each of them is a single chopstick. The philosophers are holding one of their famous _symposia_, a party for eathing and thinking and drinking. Each philosopher will alternate between thinking and eating. To eat, a philosopher needs to have hold of _both_ chopsticks. Crucially, this means that not all of the philosophers can eat at the same time.
+
+ ![Dining philosophers](dining_philosophers2.png)
+
+A fuller explanation of the Dining Philosphers problem can be found on [Wikipedia](https://en.wikipedia.org/wiki/Dining_philosophers_problem).
+
+## Deadlock
+It is easy for the philosophers to get into what is called a deadlocked state. In the analogy, each philosopher is determined to eat, and picks up the chopstick to their right, and is unwilling to put it down until after they have eaten. They are each left holding exactly one chopstick each, and are also left hungry, each unwilling to put down the chopstick the have hold of.
```

Comment:
Typo: eathing

Reply:
also suggest to remove drinking as not referenced in example

Reply:
typo: the have hold of

Reply:
Symposium specifically means drinking together,  but point taken.